src/PostgREST/Error.hs

```diff
@@ -359,11 +361,24 @@ type Authenticated = Bool
 instance PgrstError PgError where
   status (PgError authed usageError) = pgErrorStatus authed usageError
 
+  headers (PgError _ (SQL.SessionUsageError (SQL.QueryError _ _ (SQL.ResultError (SQL.ServerError "PGRST" m d _ _p))))) =
+    let
+      intoHeader (k,v) = ((CI.mk . T.encodeUtf8) k, T.encodeUtf8 v)
+    in
+      case parseMessage m of -- make sure both parses successfully
+        Just _ ->
+          case parseDetails d of
+            Just obj -> headers JSONParseError
+                          ++ map intoHeader (M.toList $ getHeaders obj)
+            Nothing -> headers JSONParseError
+        Nothing -> headers JSONParseError
```

Comment:
This could be refactored by using a Tuple:

```suggestion
    case (parseMessage m, parseDetails d) of -- make sure both parses successfully
      (Just _, Just obj) -> headers JSONParseError ++ map intoHeader (M.toList $ getHeaders obj)
      _                  -> headers JSONParseError
    where
      intoHeader (k,v) = ((CI.mk . T.encodeUtf8) k, T.encodeUtf8 v)
```

Or if you have something better, go for it. It could be applied to similar code below.

Reply:
Great suggestion! Much clearer and shorter.